packages/components/src/playground/button.tsx

```diff
@@ -0,0 +1,33 @@
+import React from "react";
+import styled from "styled-components";
+
+// Temporary props for button to test out css options for variants
+type ButtonProps = React.DetailedHTMLProps<
+  React.ButtonHTMLAttributes<HTMLButtonElement>,
+  HTMLButtonElement
+> & {
+  ref?: React.Ref<HTMLButtonElement>;
```

Comment:
Adding this to resolve a TS error where there were two conflicting declarations of `ref`?

Reply:
This is pretty interesting. I think you're right: the `ref` type coming from `React.DetailedHTMLProps` seems to be conflicting with what we're getting from `styled.button`

Specifically, it seems like:
- `React.DetailedHTMLProps` is expecting "legacy" refs (object, callback, and the deprecated string ref)
- `styled.button` is expecting non-legacy refs (object and callback refs only)

(Verified by clicking into each definition. There are many layers to go through, but they're there)

This raises an interesting issue, though: we can't use `ref` as a prop like this 😬 .

A `ref` is handled separately from normal props (kind of hard to see, but [note here that "reserved" props [which ref is, along with key] are not copied over to the props object](https://github.com/facebook/react/blob/daf38ecdfcda32a1050231fcc6dea364f8b102e3/packages/react/src/ReactElement.js#L375-L380)).

That's why normally we see something like:
- Use a name other than `ref`, such as `buttonRef`
- Use `React.forwardRef`, which adds `ref` to custom components

This case doesn't seem like we quite need `ref` exposed, yet, and I'd recommend dropping the `DetailedHTMLProps`:

```ts
type Props = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: "primary" | "secondary";
}
```

Let me know if that makes any sense at all. Also happy to talk this through in person, if that's more helpful.

---

Unrelated, but this also highlighted that I think we should have an explicit dependency on `@types/react` somewhere. This works because other dependencies are installing it, but (if we haven't yet), we'll want our own dependency on it.

Reply:
Actually, the example I linked to in the React source may not be a good example of why we can't use `ref` in this way.

A better reference is probably https://reactjs.org/docs/forwarding-refs.html, which says

> Regular function or class components don’t receive the ref argument, and ref is not available in props either.

Reply:
🤯 wow, thanks for digging into that! I agree we don't need `ref` here, since it's just a test component. When we build the actual Button, might need to revisit. And I'll add `@types/react` 👍